DQM/EcalMonitorTasks/interface/GpuTask.h

```diff
@@ -0,0 +1,106 @@
+#ifndef DQM_EcalMonitorTasks_GpuTask_H
+#define DQM_EcalMonitorTasks_GpuTask_H
+
+#include "DQWorkerTask.h"
+
+#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+
+namespace ecaldqm {
+
+  class GpuTask : public DQWorkerTask {
+  public:
+    GpuTask();
+    ~GpuTask() override {}
+
+    void addDependencies(DependencySet&) override;
+
+    bool filterRunType(short const*) override;
+
+    void beginEvent(edm::Event const&, edm::EventSetup const&, bool const&, bool&) override;
+    bool analyze(void const*, Collections) override;
+
+    template <typename DigiCollection>
+    void runOnCpuDigis(DigiCollection const&, Collections);
+    template <typename DigiCollection>
+    void runOnGpuDigis(DigiCollection const&, Collections);
+    void runOnCpuUncalibRecHits(EcalUncalibratedRecHitCollection const&, Collections);
+    void runOnGpuUncalibRecHits(EcalUncalibratedRecHitCollection const&, Collections);
+    void runOnCpuRecHits(EcalRecHitCollection const&, Collections);
+    void runOnGpuRecHits(EcalRecHitCollection const&, Collections);
+
+  private:
+    void setParams(edm::ParameterSet const&) override;
+
+    bool runGpuTask_;
+    bool gpuOnlyPlots_;
+    std::vector<int> uncalibOOTAmps_;
+
+    // Static cast to EB/EEDigiCollection when using
+    // Defined as void pointers to make compiler happy
+    void const* EBCpuDigis_;
+    void const* EECpuDigis_;
```

Comment:
I did that initially and got a compiler complaint at these lines:
https://github.com/cms-sw/cmssw/blob/c26b2fcd7b0f061bc73147b51aad9465d503d282/DQM/EcalMonitorTasks/src/GpuTask.cc#L120-L125

`EBDigiCollection` and `EEDigiCollection` cannot be `static_cast` into each other. The compiler tries to protect from the situation where the function argument `_cpuDigis` is one digi flavor and cast into the other. Doing the following will also result in a compiler error saying you can't cast EB/EE to EE/EB:
```
    if (iSubdet == EcalBarrel)
      EBCpuDigis_ = static_cast<EBDigiCollection const *>(&_cpuDigis);
    else
      EECpuDigis_ = static_cast<EBDigiCollection const *>(&_cpuDigis);
```

The compiler appears to learn that the template type `DigiCollection` can be either flavor from the calls function calls here:
https://github.com/cms-sw/cmssw/blob/c26b2fcd7b0f061bc73147b51aad9465d503d282/DQM/EcalMonitorTasks/interface/GpuTask.h#L53-L62
and tries to ensure that every line can execute for either type.

Reply:
Why would you do `EECpuDigis_ = static_cast<EBDigiCollection const *>(&_cpuDigis);` and not `EECpuDigis_ = static_cast<EEDigiCollection const *>(&_cpuDigis);`?

Reply:
The compiler does not "learn" anything, it simply tries to compile the code as it is written.

When you call `GpuTask::runOnCpuDigis()` passing an `EBDigiCollection` as the first argument, the whole function is compiled substituting `EBDigiCollection` in place of the template type `DigiCollection`, and becomes something like
```c++
  void GpuTask::runOnCpuDigis(EBDigiCollection const& _cpuDigis, Collections _collection) {
    MESet& meDigiCpu(MEs_.at("DigiCpu"));
    MESet& meDigiCpuAmplitude(MEs_.at("DigiCpuAmplitude"));

    int iSubdet(_collection == kEBCpuDigi ? EcalBarrel : EcalEndcap);

    // Save CpuDigis for comparison with GpuDigis
    // Static cast to EB/EEDigiCollection during use
    // Stored as void pointers to make compiler happy
    if (iSubdet == EcalBarrel)
      EBCpuDigis_ = &_cpuDigis;
    else
      EECpuDigis_ = &_cpuDigis;
  ...
```
so both assignment `EBCpuDigis_ = &_cpuDigis` and `EECpuDigis_ = &_cpuDigis` must be valid.

If one is of type `EBDigiCollection *` and the other is of type `EEDigiCollection *`, it cannot compile - it doesn't matter that only one of the branches would be executed.

One way to fix the code could be to conditionally compile one of the two assignments, depending on the type of the digi collection:
```c++
    if constexpr (std::is_same_v<DigiCollection, EBDigiCollection>) {
      assert(iSubdet == EcalBarrel);
      EBCpuDigis_ = &_cpuDigis;
    } else {
      assert(iSubdet == EcalEndcap);
      EECpuDigis_ = &_cpuDigis;
    }
```

The use of `if constexpr` instead of a plain `if` tells the compiler to evaluate the condition at compile time instead of runtime, and compile only the corresponding branch.

Reply:
> Why would you do `EECpuDigis_ = static_cast<EBDigiCollection const *>(&_cpuDigis);` and not `EECpuDigis_ = static_cast<EEDigiCollection const *>(&_cpuDigis);`?

@thomreis Sorry that was a typo in the comment. I did try it with the correct cast.

@fwyzard Thank you for the clarification and suggestion! I'll implement this along with the other suggestions.